grid2op/multi_agent/multiAgentEnv.py

```diff
@@ -0,0 +1,562 @@
+            tmp_subgrid.mask_storage
+        ]
+        tmp_subgrid.storage_charging_efficiency = cent_env_cls.storage_charging_efficiency[
+            tmp_subgrid.mask_storage
+        ]
+        tmp_subgrid.storage_discharging_efficiency = cent_env_cls.storage_discharging_efficiency[
+            tmp_subgrid.mask_storage
+        ]
+
+        # grid layout
+        tmp_subgrid.grid_layout = None if cent_env_cls.grid_layout is None else {
+            k : cent_env_cls.grid_layout[k]
+            for k in tmp_subgrid.name_sub
+        }
+
+        # alarms
+        tmp_subgrid.dim_alarms = 0
+        tmp_subgrid.alarms_area_names = []
+        tmp_subgrid.alarms_lines_area = {}
+        tmp_subgrid.alarms_area_lines = []
+        if cent_env_cls.dim_alarms != 0:
+            warnings.warn("Alarms are not yet handled by the \"multi agent\" environment. They have been deactivated")
+        
+        extra_name = domain["agent_name"]
+        if self._add_to_name is not None:
+            extra_name += f"{self._add_to_name}"
+        res_cls = SubGridObjects.init_grid(gridobj=tmp_subgrid, extra_name=extra_name)
+        res_cls.assert_grid_correct_cls()
+        return res_cls
+    
+    def _build_observation_spaces(self):
+        """Build observation spaces from given domains for each agent
+
+        Args:
+            global_obs (bool, optional): The agent can observe the whole grid. Defaults to True.
+
+        Raises:
+            NotImplementedError: Local observations are not implemented yet
+        """
+        
+        if self._observation_domains is None:
+            for agent in self.agents:
+                self.observation_spaces[agent] = self._cent_env.observation_space.copy()
+        else:
+            raise NotImplementedError("Local observations are not available yet !")
+    
+    
+    def _build_action_spaces(self):
+        """Build action spaces from given domains for each agent
+        The action class is the same as 
+        """
+        for agent in self.agents: 
+            _cls_agent_action_space = ActionSpace.init_grid(gridobj=self._subgrids_cls['action'][agent], extra_name=agent)
+            self.action_spaces[agent] = _cls_agent_action_space(
+                gridobj=self._subgrids_cls['action'][agent],
+                actionClass=self._cent_env._actionClass,
+                legal_action=self._cent_env._game_rules.legal_action,
+            )
+    
+    
+    def _update_observations(self, _update_state = True):
+        """Update agents' observations from the global observation given by the self._cent_env
+
+        Args:
+            observation (BaseObservation): _description_
+        """
+        #for agent in self.agents:
+        #    self.observations[agent] = observation.copy()
+        
+        if self._cent_env.__closed:
+            raise EnvError("This environment is closed. You cannot use it anymore.")
+        if not self.__is_init:
```

Comment:
missing a "self._cent_env.__is_init" no ?

Beside, trust me, these two checks will NOT work, so remove them and put a `TODO BEN: check that cent env is initialized and not closed"

And in general, don't do "half of stuff". When you do them, you commit to doing them completely. When you don't do them, you write "todos" and "raise NotImplementedError(...) to make sure you remember there is something to do at this stage of the code

Reply:
Ok done